engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java

```diff
@@ -134,13 +145,105 @@ Object doPanicSentinel(
     throw self;
   }
 
-  @Specialization
+  /**
+   * Resolves symbol to a Warning method, if possible.
+   *
+   * <p>A regular method dispatch logic will extract/append warnings of `self` before invoking the
+   * actual method dispatch logic. This allows for ignoring complexity related to the presence of
+   * warnings but prevents us from manipulating warnings directly in the Enso code (they have just
+   * been removed). `resolveWarningFunction` will attempt to resolve the symbol in the Warning type
+   * scope, if possible. Additionally, we check if under the non-warning `self`, the symbol would
+   * resolve to `Any`. If not, it means that we should employ a regular method dispatch logic, due
+   * to a method name clash. E.g. if some collection type Foo defines a `has_warnings` method, we
+   * should dispatch the call to `Foo`'s `has_warning` rather than to a `Warning` or `Any`'s `one.
+   *
+   * @param self `self` argument that has some warnings
+   * @param symbol symbol to be resolved
+   * @param types TypesLibrary instance
+   * @param warnings WarningsLibrary instance
+   * @return resolved Warning method to be called
+   */
+  Function resolveWarningFunction(
+      Object self, UnresolvedSymbol symbol, TypesLibrary types, WarningsLibrary warnings) {
+    Object selfWithoutWarnings;
+    try {
+      selfWithoutWarnings = warnings.removeWarnings(self);
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+
+    Type typeOfSymbol = symbol.resolveTypeFor(types.getType(selfWithoutWarnings));
+    Builtins builtins = EnsoContext.get(this).getBuiltins();
+    if (typeOfSymbol == builtins.any()) {
+      return symbol
+          .getScope()
+          .lookupMethodDefinition(builtins.warning().getEigentype(), symbol.getName());
+    }
+
+    return null;
+  }
+
+  Object[] argumentsWithExplicitSelf(FunctionSchema cachedSchema, Object[] arguments) {
+    Object[] arguments1;
+    if (!cachedSchema.isFullyApplied()) {
+      arguments1 = new Object[cachedSchema.getArgumentsCount()];
+      System.arraycopy(arguments, 0, arguments1, 1, arguments.length);
+      arguments1[0] = arguments[0];
+    } else {
+      arguments1 = arguments;
+    }
+    return arguments1;
+  }
+
+  CallArgumentInfo[] callArgumentsWithExplicitSelf() {
+    int length = invokeFunctionNode.getSchema().length;
```

Comment:
Caching call info based on `invokeFunctionNode` field? That is very suspicious. How that is supposed to work?